test: atomic: multiple equal priority threads access atomic variable

yingmingx opened the pull request:
This is a testcase for testing if two preempt child threads with equal priority 
and time slice 1ms can atomiclly access the same atomic value. 

Signed-off-by: Ying ming <mingx.ying@intel.com>

LeiW000 commented on tests/kernel/atomic/src/main.c:
```diff
@@ -0,0 +1,71 @@
+	k_sched_time_slice_set(0, 10);
+
+	printk("total_atomic = %d\n", total_atomic);
+	zassert_true(total_atomic == (TEST_CYCLE * THREADS_NUM), "NULL");
```
Should we add an error message for the failure of zassert_true(), rather than "NULL"?

andyross replied:
Heh.  In fact `NULL` (the pointer value, not `"NULL"` the null-terminated string literal) is a common idiom for "no message".  This looks like a typo.  Indeed we shouldn't be logging the string "NULL" as a message describing a failure.

yingmingx replied:
Yeah，Steven mean I should leave some error message instead of writing NULL here. This is reasonable.

andyross commented on tests/kernel/atomic/src/main.c:
```diff
@@ -0,0 +1,71 @@
+void test_main(void)
+{
+	ztest_test_suite(atomic_threads,
+			ztest_1cpu_unit_test(test_threads_access_atomic));
```
Why 1cpu?  If there was ever a test you would want to exercise SMP functionality, it would be this one.

yingmingx replied:
Thank you for reminding me. Atomic operations can be used in multithreading, including SMP.

nashif reviewed (changes requested):
Can this test please be folded into existing atomic test, maybe `tests/kernel/common/src/atomic.c`

yingmingx commented:
OK. Testcase has been moved to file tests/kernel/common/src/atomic.c.

LeiW000 commented:
@nashif, the code has been moved into atomic.c as you suggested. Could you take look at the PR again? We would like to get it merged before 05/07.

nashif commented on tests/kernel/common/prj.conf:
```diff
@@ -6,3 +6,5 @@ CONFIG_BOOT_DELAY=500
 CONFIG_IRQ_OFFLOAD=y
 CONFIG_TEST_USERSPACE=y
 CONFIG_BOUNDS_CHECK_BYPASS_MITIGATION=y
+CONFIG_TIMESLICING=y
```
isnt this enabled by default?

yingmingx replied:
I check this in the zephyr doc and it is by default. Later, I'll check this doc when enable new configurations. Thanks for reminding.

yingmingx replied:
I've removed this configuration in code.

nashif commented on tests/kernel/common/src/atomic.c:
```diff
@@ -267,6 +279,49 @@ void test_atomic(void)
+	k_tid_t tid[THREADS_NUM];
+
+	/* enable time slice 1ms at priority 10 */
+	k_sched_time_slice_set(1, K_PRIO_PREEMPT(10));
```
why is time slicing relevant here?

yingmingx replied:
Although CONFIG_TIMESLICING is by default, the default CONFIG_TIMESLICE_SIZE is 0. A time slice size of zero means "no limit" (i.e. an infinitely large time slice).
So I use function k_sched_time_slice_set to set the time slice at a very small value, it make the preempt thread which have equal thread priority can be scheduling by timeslice.

yingmingx replied:
@nashif The previous description is not clear. The testcase description is update to "This is a testcase for testing if two preempt child threads with equal priority and time slice 1ms can atomiclly access the same atomic value." The 2 preempt child threads with equal priority can not be preempted with default timeslice 0. That's why I need set time slice to 1ms.
Hope this clarifies.

Commit:
test: atomic: multiple equal priority threads access atomic variable

Add an testcase. Creat two preempt threads with equal priority to
atomiclly access the same atomic value. Because these preempt
threads are of equal priority, so enable time slice to make
them scheduled. The thread will execute for some time.
In this time, the two sub threads will be scheduled separately
according to the time slice.

Signed-off-by: Ying ming <mingx.ying@intel.com>
